File: annotation_software/crop_generator/cropgenerator/generatorobjects/generatorobjects.py
# Class definition for objects used in the crop_generator module and database
# Author: Michael B. Lance
# created: April 4, 2025
# updated: August, 2 2025
#---------------------------------------------------------------------------------------------------------------------------#

import numpy as np
import cv2
import os
from flask.json.provider import DefaultJSONProvider
import cv2
from abc import ABC, abstractmethod
from datetime import datetime
import io
import PIL.Image as PillowImage
from dataclasses import dataclass
from uuid import UUID
from flask_login import UserMixin

#---------------------------------------------------------------------------------------------------------------------------#
# ABC for easy serialization of child classes

class CgOBJ(ABC):
	@abstractmethod
	def serialize(self) -> dict:
		pass
	
#---------------------------------------------------------------------------------------------------------------------------#
# Project Management -- For Database use only

@dataclass
class Project(CgOBJ):
	''' Dataclass for representing projects from the databse
	
	'''
	project_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class Schema(CgOBJ):
	''' Dataclass for representing scehmas from the databse
	
	'''
	schema_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class Label(CgOBJ):
	''' Dataclass for representing labels from the databse
	
	'''
	label_id: int
	schema_id: int
	label: int
	name: str
	image_link: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'label': self.label,
			'name': self.name,
			'image_link': self.image_link,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class HerdUnit(CgOBJ):
	''' Dataclass for representing Herd Units from the database
	
	'''
	herd_unit_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class Model(CgOBJ):
	''' Dataclass for representing Models from the database
	
	'''
	model_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}
	
@dataclass
class Survey(CgOBJ):
	''' Dataclass for representing Surveys from the database
	
	'''
	survey_id: int
	survey_year: int
	name: str
	additional_info: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'survey_year': self.survey_year,
			'name': self.name,
			'additional_info': self.additional_info,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}
	
#---------------------------------------------------------------------------------------------------------------------------#
# User Management -- For Database use only

class User(UserMixin, CgOBJ):
	def __init__(self, user_id: int, username: str, external_auth_id: str, external_auth_provider: str, status: str,
				 created: datetime.date, modified: datetime.date, last_login: datetime,  locale: str, uuid: UUID, roles: tuple[str] | None = None):
		self.id = str(uuid) # this is this way to make Flask-Login happy
		self.user_id = user_id
		self.username = username
		self.external_auth_id = external_auth_id
		self.external_auth_provider = external_auth_provider
		self.status = status
		self.created = created
		self.modified = modified
		self.last_login = last_login
		self.locale = locale
		self.uuid = uuid
		self.roles = roles
	
	def get_id(self) -> str:
		return self.id
	
	def has_role(self, role_name: str) -> bool:
		return role_name in self.roles

	def serialize(self):
		return {
			'username': self.username,
			'status': self.status,
			'created': self.created,
			'modified': self.modified,
			'last_login': self.last_login,
			'locale': self.locale,
			'uuid': self.id,
			'roles': self.roles
		}
	
@dataclass
class Role(CgOBJ):
	role_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}
	
@dataclass
class Organization(CgOBJ):
	organization_id: int
	name: str
	created: datetime
	modified: datetime
	logo_url: str | None
	uuid: UUID

	def serialize(self):
		return {
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'logo_url': self.logo_url,
			'uuid': self.uuid
		}

#---------------------------------------------------------------------------------------------------------------------------#
# Core 

class Box(CgOBJ):
	''' A Box contains dimensional data for crops and predictions
	
	'''
	def __init__(self, top_left: tuple[int]=None, bottom_right: tuple[int]=None):
		self.top_left = top_left
		self.bottom_right = bottom_right

	def get_center(self) -> tuple:
		x = np.mean([self.top_left[0], self.bottom_right[0]])
		y = np.mean([self.top_left[1], self.bottom_right[1]])

		return ((abs(x), abs(y)))

	def get_points(self) -> list[int]:
		return [self.top_left[0], self.top_left[1], self.bottom_right[0], self.bottom_right[1]]

	def calc_iou(self, box_2) -> float:
		# Slightly modified from https://machinelearningspace.com/intersection-over-union-iou-a-comprehensive-guide/
		#Extract bounding boxes coordinates
		x0_A, y0_A, x1_A, y1_A = self.get_points()
		x0_B, y0_B, x1_B, y1_B = box_2.get_points()
		
		# Get the coordinates of the intersection rectangle
		x0_I = max(x0_A, x0_B)
		y0_I = max(y0_A, y0_B)
		x1_I = min(x1_A, x1_B)
		y1_I = min(y1_A, y1_B)
		#Calculate width and height of the intersection area.
		width_I = x1_I - x0_I 
		height_I = y1_I - y0_I

		# Handle the negative value width or height of the intersection area
		width_I = 0 if width_I < 0 else width_I
		height_I = 0 if height_I < 0 else height_I
		# Calculate the intersection area:
		intersection = width_I * height_I
		# Calculate the union area:
		width_A, height_A = x1_A - x0_A, y1_A - y0_A
		width_B, height_B = x1_B - x0_B, y1_B - y0_B
		union = (width_A * height_A) + (width_B * height_B) - intersection
		# Calculate the IoU:
		return intersection/union
	
	def serialize(self) -> dict:
		return {
			'top_left': list(self.top_left),
			'bottom_right': list(self.bottom_right),
		}

@dataclass
class Image(CgOBJ):
	def __init__(self, image_id: int, herd_unit_id: int, survey_id: int, name: str,
				in_training: bool, crops_generated: int, reviewed_by_user_id: int, opened_by_user_id: int,
				created: datetime, modified: datetime, image_length_px: int, image_width_px: int,
				uuid: UUID):
		self.image_id = image_id
		self.herd_unit_id = herd_unit_id
		self.survey_id = survey_id
		self.name = name
		self.in_training = in_training
		self.crops_generated = crops_generated
		self.reviewed_by_user_id = reviewed_by_user_id
		self.opened_by_user_id = opened_by_user_id
		self.created = created
		self.modified = modified
		self.image_length_px = image_length_px
		self.image_width_px = image_width_px
		self.uuid = uuid
		self.image = None

	def set_image(self, image_data: np.ndarray):
		if isinstance(image_data, bytes):
			try:
				pil_image = PillowImage.open(io.BytesIO(image_data))
				pil_image = pil_image.convert('RGB')
				self.image = cv2.cvtColor(np.array(pil_image), cv2.COLOR_RGB2BGR)
			except Exception as e:
				print(f"Error converting bytes to image: {e}")
				self.image = None
		elif isinstance(image_data, np.ndarray):
			self.image = image_data
		else:
			print(f"Unsupported type: {type(image_data)}")
			self.image = None

	def get_image(self) -> np.ndarray:
		if self.image is not None:
			return self.image
		else:
			self.image = cv2.imread(os.path.join(f'{self.local_path}', f'{self.name}.JPG'))
			return self.image

	def delete_image(self):
		del self.image

	def serve(self, img_format: str):
		_, self.img_encoded = cv2.imencode(img_format, self.get_image())
		return self.img_encoded.tobytes()
		
	def serialize(self) -> dict:
		return {
			'name': self.name,
			'in_training': self.in_training,
			'crops_generated': self.crops_generated,
			'created': self.created,
			'modified': self.modified,
			'image_length_px': self.image_length_px,
			'image_width_px': self.image_width_px,
			'uuid': self.uuid
		}

class Prediction(CgOBJ):
	def __init__(self, pred_id: int, image_id: int, model_id: int, score: float, box_tx: int, box_ty: int,
				 box_bx: int, box_by: int, created: datetime | None, modified: datetime | None, uuid: UUID):
		self.pred_id = pred_id
		self.image_id = image_id
		self.model_id = model_id
		self.score = score
		self.dimensions = Box((box_tx, box_ty), (box_bx, box_by))
		self.created = created 
		self.modified = modified
		self.uuid = uuid
	
	def serialize(self) -> dict:
		return {
			'id': self.id,
			'model': self.model.serialize(),
			'dimensions': self.dimensions.serialize(),
			'score': self.score,
			'label': self.label,
		}

class Crop(Image):
	def __init__(self, db_id: int=None, image_id: int=None, name: str=None, dimensions: Box=None):
		super().__init__(db_id, name)
		self.image_id = image_id
		self.crop_dimensions = dimensions

	def calc_iou(self, box):
		return self.crop_dimensions.calc_iou(box)
	
	def serialize(self) -> dict:
		return {
			'id': self.id,
			'image_id': self.image_id,
			'name': self.name,
			'dimensions': self.crop_dimensions.serialize(),
			'url' : self.url,
		}
	
class PredictionCrop(Crop):
	def __init__(self, pred_crop_id: int=None, image_id: int=None, name: str=None, score: float=None, label: int=None, dimensions: Box=None, 
				 url: str=None):
		self.pred_crop_id = pred_crop_id
		self.image_id = image_id
		self.name = name
		self.score = score
		self.label = label 
		self.dimensions = dimensions
		self.approved = False
	
	def serialize(self) -> dict:
		return {
			'id': self.pred_crop_id,
			'image_id': self.image_id,
			'name': self.name,
			'score': self.score,
			'label': self.label,
			'dimensions': self.dimensions.serialize(),
			'approved': self.approved,
		}
#---------------------------------------------------------------------------------------------------------------------------#

class CropgenJSONPRovider(DefaultJSONProvider):
	def default(self, obj):
		if isinstance(obj, CgOBJ):
			return obj.serialize()
		return super().default(obj)


#---------------------------------------------------------------------------------------------------------------------------#

File: annotation_software/generator_web_app/src/types/generatorobjects.ts
// Typescript Class definition analogues for crop_generator objects 
// Author: Michael B. Lance
// Created: April 9, 2025
// Updated: August 8, 2025
//---------------------------------------------------------------------------------------------------------------------------//

import { mean } from 'lodash';

//---------------------------------------------------------------------------------------------------------------------------//

export interface Box_intf {
	top_left: number[];
	bottom_right: number[];
}

export class Box implements Box_intf {
    top_left: number[];
    bottom_right: number[];
    constructor (box: Box_intf) {
        this.top_left = box.top_left;
        this.bottom_right = box.bottom_right;
    }
    get_center(): number[] {
        let x = mean([this.top_left[0], this.bottom_right[0]]);
        let y = mean([this.top_left[1], this.bottom_right[1]]);
        return [x, y];
    }
    get_points(): number[] {
        return [this.top_left[0], this.top_left[1], this.bottom_right[0], this.bottom_right[1]];
    }
	get_width(): number {
		return Math.abs(this.top_left[0] - this.bottom_right[0]);
	}
	get_height(): number {
		return Math.abs(this.top_left[1] - this.bottom_right[1]);
	}

}

//---------------------------------------------------------------------------------------------------------------------------//

export interface HerdUnit_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class HerdUnit implements HerdUnit_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    
    constructor(herdunit: HerdUnit_intf) {
        this.name = herdunit.name;
        this.created = new Date(herdunit.created);
        this.modified = new Date(herdunit.modified);
        this.uuid = herdunit.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Image_intf {
	name: string;
	in_training: boolean;
	crops_generated: number;
	created: Date;
	modified: Date;
	image_length_px: number;
	image_width_px: number;
	uuid: string;
}

export class Image implements Image_intf {
    name: string;
	in_training: boolean;
	crops_generated: number;
	created: Date;
	modified: Date;
	image_length_px: number;
	image_width_px: number;
	uuid: string;
  
    constructor(img: Image_intf) {
        this.name = img.name;
		this.in_training = img.in_training;
		this.crops_generated = img.crops_generated;
		this.created = new Date(img.created);
		this.modified = new Date(img.modified);
		this.image_length_px = img.image_length_px;
		this.image_width_px = img.image_width_px;
		this.uuid = img.uuid
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Prediction_intf { 
    dimensions: Box;
    score: number;
    label: number;
	created: Date;
	modified: Date;
	uuid: string;
}

export class Prediction implements Prediction_intf {
    dimensions: Box;
    score: number;
    label: number;
	created: Date;
	modified: Date;
	uuid: string;

    constructor(pred: Prediction_intf) {
        this.dimensions = new Box(pred.dimensions);
        this.score = pred.score;
        this.label = pred.label;    
		this.created = new Date(pred.created);
		this.modified = new Date(pred.modified);
		this.uuid = pred.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Annotation_intf {
	dimensions: Box;
	score: number;
    label: number;
	created: Date;
	modified: Date;
	uuid: string;
}

export class Annotation implements Annotation_intf {
	dimensions: Box;
	score: number;
    label: number;
	created: Date;
	modified: Date;
	uuid: string;

	constructor (annotation: Annotation_intf) {
		this.dimensions = new Box(annotation.dimensions);
        this.score = annotation.score;
        this.label = annotation.label;    
		this.created = new Date(annotation.created);
		this.modified = new Date(annotation.modified);
		this.uuid = annotation.uuid;
	}
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Crop_intf {
    id: number,
    image_id: number,
    name: string,
    dimensions: Box,
    url: string,
}

export class Crop implements Crop_intf {
    id: number;
    image_id: number;
    name: string;
    dimensions: Box;
    url: string;

    constructor(crop: Crop_intf) {
        
        this.id = crop.id;
        this.image_id = crop.image_id;
        this.name = crop.name;
        this.dimensions = crop.dimensions;
        this.url = crop.url;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface PredictionCrop_intf {
    image_id: string;
    name: string;
    score: number;
    label: number;
    dimensions: Box;
	bounding_box: Box;
    approved: boolean;
	uuid: string;
}

export class PredictionCrop implements PredictionCrop_intf {
    image_id: string;
    name: string;
    score: number;
    label: number;
    approved: boolean;
    dimensions: Box;
	bounding_box: Box;
    url: string;
	uuid: string;
	draw_box: boolean = false;
    
    constructor(predcrop: PredictionCrop_intf, url: string) { 
        this.image_id = predcrop.image_id;
        this.name = predcrop.name;
        this.score = predcrop.score;
        this.label = predcrop.label;
        this.dimensions = new Box(predcrop.dimensions);
		this.bounding_box = new Box(predcrop.bounding_box);
        this.approved = predcrop.approved;
		this.uuid = predcrop.uuid;
        this.url = url;
    }
}
//---------------------------------------------------------------------------------------------------------------------------//

export interface Role_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Role implements Role_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(role: Role_intf) {
        this.name = role.name;
        this.created = new Date(role.created);
        this.modified = new Date(role.modified);
        this.uuid = role.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface User_intf {
    username: string;
    status: string;
    created: Date;
    modified: Date;
    last_login: Date;
    locale: string;
    uuid: string;
    roles: Role_intf[];
}

export class User {
    username: string;
    status: string;
    created: Date;
    modified: Date;
    last_login: Date;
    locale: string;
    roles: Role[];
    uuid: string;

    constructor(usr: User_intf) {
        this.username = usr.username;
        this.status = usr.status;
        this.created = new Date(usr.created);
        this.modified = new Date(usr.modified);
        this.last_login = new Date(usr.last_login);
        this.locale = usr.locale;
        this.roles = [];
        for (const role of usr.roles) this.roles.push(new Role(role));
        this.uuid = usr.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Organization_intf {
    name: string;
    created: Date;
    modified: Date;
    logo_url: string | undefined;
    uuid: string;
}

export class Organization implements Organization_intf {
    name: string;
    created: Date;
    modified: Date;
    logo_url: string | undefined;
    uuid: string;

    constructor(Org: Organization_intf) {
        this.name = Org.name;
        this.created = new Date(Org.created);
        this.modified = new Date(Org.modified);
        this.logo_url = Org.logo_url;
        this.uuid = Org.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Project_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Project implements Project_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(Proj: Project_intf) {
        this.name = Proj.name;
        this.created = new Date(Proj.created);
        this.modified = new Date(Proj.modified);
        this.uuid = Proj.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Schema_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Schema implements Schema_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(Schem: Schema_intf) {
        this.name = Schem.name;
        this.created = new Date(Schem.created);
        this.modified = new Date(Schem.modified);
        this.uuid = Schem.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Label_intf {
    label: number;
    name: string;
    image_link: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Label implements Label_intf {
    label: number;
    name: string;
    image_link: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(lbl: Label_intf) {
        this.label = lbl.label;
        this.name = lbl.name;
        this.image_link = lbl.image_link;
        this.created = new Date(lbl.created);
        this.modified = new Date(lbl.modified);
        this.uuid = lbl.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Survey_intf {
    survey_year: number;
    name: string;
    additional_info: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Survey implements Survey_intf {
    survey_year: number;
    name: string;
    additional_info: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(srvy: Survey_intf) {
        this.survey_year = srvy.survey_year;
        this.name = srvy.name;
        this.additional_info = srvy.additional_info;
        this.created = new Date(srvy.created);
        this.modified = new Date(srvy.modified);
        this.uuid = srvy.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface Model_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Model implements Model_intf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
    constructor(mdl: Model_intf) {
        this.name = mdl.name;
        this.created = new Date(mdl.created);
        this.modified = new Date(mdl.modified);
        this.uuid = mdl.uuid
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export default {};